File: typescript/examples/string-slugify.test.ts
// カード string-slugify の Contract を検証するテスト
import { deburr, kebabCase } from 'es-toolkit';
import { describe, expect, it } from 'vitest';

const slugify = (s: string) => kebabCase(deburr(s));

describe('string-slugify: kebabCase', () => {
  it('Usage のとおりアクセントを除去し、小文字・ハイフン区切りにする', () => {
    expect(slugify('Crème Brûlée à la Mode!')).toBe('creme-brulee-a-la-mode');
    expect(slugify('Hello, World 2024')).toBe('hello-world-2024');
  });

  it('deburr は結合記号を除き、ラテン拡張文字を ASCII に置き換える', () => {
    expect(deburr('Crème brûlée')).toBe('Creme brulee');
    expect(deburr('é')).toBe('e');
    expect(deburr('Ærøskøbing')).toBe('Aeroskobing');
    expect(deburr('Straße')).toBe('Strasse');
    expect(deburr('Łódź')).toBe('Lodz');
  });

  it('連続する区切りや先頭・末尾の記号でハイフンが重ならない', () => {
    expect(slugify('  --Hello,,  World--  ')).toBe('hello-world');
  });

  it('数字は独立した単語になる', () => {
    expect(slugify('Release 2024')).toBe('release-2024');
    expect(slugify('web3 Rock & Roll')).toBe('web-3-rock-roll');
  });

  it('ラテン文字以外は変換も除去もされず残る', () => {
    expect(deburr('日本語')).toBe('日本語');
    expect(slugify('こんにちは 世界')).toBe('こんにちは-世界');
    expect(slugify('日本語のTitle 2024')).toBe('日本語のtitle-2024');
  });

  it('絵文字は 1 単語として残る', () => {
    expect(slugify('hello 🚀 world')).toBe('hello-🚀-world');
  });

  it('空文字や記号のみは空文字を返し、例外を投げない', () => {
    expect(slugify('')).toBe('');
    expect(slugify('!!! ???')).toBe('');
    expect(() => slugify("Don't & Rock")).not.toThrow();
    expect(slugify("Don't stop")).toBe('don-t-stop');
  });

  it('Pitfalls の ASCII 化の前処理で非 ASCII を落とせる', () => {
    const ascii = (s: string) => kebabCase(deburr(s).replace(/[^\x00-\x7F]/g, ' '));
    expect(ascii('Crème こんにちは 世界')).toBe('creme');
    expect(ascii('こんにちは')).toBe('');
  });
  it('単独の絵文字は残り、ZWJ 結合列は要素ごとに分かれる', () => {
    expect(kebabCase(deburr('🐶'))).toBe('🐶');
    expect(kebabCase(deburr('👨‍👩‍👧'))).toBe('👨-👩-👧');
  });
});

File: python/examples/string-slugify_test.py
"""string-slugify: python-slugify の slugify の Contract を検証する。"""

import re

import pytest
from slugify import slugify

ASCII_SLUG = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*$")


def test_deburrs_and_kebab_cases() -> None:
    """アクセントを除去し、小文字・ハイフン区切りにする。"""
    assert slugify("Crème Brûlée à la Mode!") == "creme-brulee-a-la-mode"
    assert slugify("Hello, World 2024") == "hello-world-2024"
    assert slugify("naïve café") == "naive-cafe"


def test_separators_do_not_repeat_or_trail() -> None:
    """区切りが連続しても重ならず、先頭・末尾にも付かない。"""
    assert slugify("a--b__c") == "a-b-c"
    assert slugify("  multiple   spaces  ") == "multiple-spaces"
    assert slugify("Hello, World 2024!!") == "hello-world-2024"


def test_latin_extended_letters_are_transliterated() -> None:
    """ラテン拡張文字は ASCII に転写される。"""
    assert slugify("Æther Straße Øre Łódź") == "aether-strasse-ore-lodz"


def test_japanese_and_chinese_are_transliterated_not_removed() -> None:
    """日本語・中国語は除去ではなく転写される（転写結果はバックエンド依存なので ASCII であることと接頭だけを検証）。"""
    ja = slugify("こんにちは 世界")
    assert ASCII_SLUG.match(ja) and ja.startswith("konni") and "-" in ja
    assert slugify("北京") in {"bei-jing", "beijing"}
    assert ASCII_SLUG.match(slugify("Hello こんにちは World"))


def test_emoji_is_removed_and_entities_are_decoded() -> None:
    """絵文字は除去され、HTML エンティティはデコードしてから処理される。"""
    assert slugify("hello 🐶 world") == "hello-world"
    assert slugify("Tom &amp; Jerry") == "tom-jerry"
    assert slugify("&#169; 2024") == "2024"


def test_digits_are_not_split_from_letters() -> None:
    """数字は前後の文字と分割されない。½ は 1-2 になる。"""
    assert slugify("web3") == "web3"
    assert slugify("userProfileURL") == "userprofileurl"
    assert slugify("½ cup") == "1-2-cup"


def test_max_length_cuts_from_the_end() -> None:
    """max_length を超える分は末尾から切り、末尾の区切りは落とす。"""
    assert slugify("Hello World Foo Bar", max_length=8) == "hello-wo"
    assert slugify("Hello World Foo Bar", max_length=6) == "hello"
    assert slugify("Hello World Foo Bar", max_length=12) == "hello-world"


def test_max_length_with_word_boundary_keeps_whole_words() -> None:
    """word_boundary=True なら完全な単語だけを残す。"""
    assert slugify("Hello World Foo Bar", max_length=8, word_boundary=True) == "hello"
    assert slugify("Hello World Foo Bar", max_length=15, word_boundary=True) == "hello-world-foo"


def test_separator_lowercase_and_allow_unicode_options() -> None:
    """separator / lowercase / allow_unicode で出力を変えられる。"""
    assert slugify("Hello World", separator="_") == "hello_world"
    assert slugify("Hello World", lowercase=False) == "Hello-World"
    assert slugify("Crème", allow_unicode=True) == "crème"
    assert slugify("こんにちは 世界", allow_unicode=True) == "こんにちは-世界"


def test_replacements_run_before_transliteration() -> None:
    """replacements で転写前に文字列置換できる。& は既定では消える。"""
    assert slugify("Rock & Roll") == "rock-roll"
    assert slugify("Rock & Roll", replacements=[["&", "and"]]) == "rock-and-roll"


def test_apostrophe_is_a_separator() -> None:
    """アポストロフィは区切りになる。"""
    assert slugify("Don't") == "don-t"


def test_empty_and_symbol_only_input() -> None:
    """空文字・記号のみは空文字を返す。"""
    assert slugify("") == ""
    assert slugify("!!!") == ""


def test_non_string_raises_type_error() -> None:
    """str / bytes / bytearray 以外は TypeError。"""
    with pytest.raises(TypeError):
        slugify(None)  # type: ignore[arg-type]
    with pytest.raises(TypeError):
        slugify(123)  # type: ignore[arg-type]


def test_is_pure() -> None:
    """同じ入力からは同じ結果が得られる。"""
    text = "Crème Brûlée"
    assert slugify(text) == slugify(text) == "creme-brulee"
    assert text == "Crème Brûlée"
